app/scripts/modules/core/src/filterModel/FilterModelService.ts

```diff
@@ -175,37 +27,91 @@ export class FilterModelService {
-      );
-
-      ReactInjector.$state.go('.', newFilters, { inherit: false });
+      const toParams = FilterModelService.mapSortFilterToRouterParams(filterModel);
+      ReactInjector.$state.go('.', toParams);
     };
 
     return filterModel;
   }
 
+  // Maps the sortFilter data from an IFilterModel object to router params
+  public static mapSortFilterToRouterParams(filterModel: IFilterModel) {
+    const { sortFilter, config } = filterModel;
+    return config.reduce((acc, filter) => ({ ...acc, [filter.param]: sortFilter[filter.model] }), {});
+  }
+
+  // Maps router param values to sortFilter values, applying known default values if the parameter is missing
+  public static mapRouterParamsToSortFilter(filterModel: IFilterModel, params: any) {
+    const getValueIfNill = (filterType: string) => {
+      switch (filterType) {
+        case 'trueKeyObject':
+          return {};
+        case 'string':
+          return '';
+        case 'boolean':
+          return false;
+        case 'inverse-boolean':
+          return true;
+      }
+      return undefined;
+    };
+
+    const iFilterConfigs = filterModel.config;
+
+    return iFilterConfigs.reduce(
+      (acc, filter) => {
+        const valueIfNil = getValueIfNill(filter.type);
+        const rawValue = params[filter.param];
+        const paramValue = isNil(rawValue) ? valueIfNil : rawValue;
+        // Clone deep so angularjs mutations happen on a different object reference
+        return { ...acc, [filter.model]: cloneDeep(paramValue) };
+      },
+      {} as any,
+    );
+  }
+
+  public static registerRouterHooks(filterModel: IFilterModel, stateGlob: string) {
+    const { transitionService } = ReactInjector.$uiRouter;
+    const filterParams = filterModel.config.map(cfg => cfg.param);
+    let savedParamsForScreen: any = {};
+
+    // When exiting the screen but staying in the app, save the filters for that screen
+    transitionService.onSuccess({ exiting: stateGlob, retained: '**.application' }, trans => {
+      const fromParams = trans.params('from');
+      savedParamsForScreen = pick(fromParams, filterParams);
+    });
+
+    // When entering the screen and staying in the app, restore the filters for that screen
+    transitionService.onBefore({ entering: stateGlob, retained: '**.application' }, trans => {
+      const toParams = trans.params();
+      const hasFilters = filterParams.some(key => !isNil(toParams[key]));
+
+      const savedParams = savedParamsForScreen;
+      const hasSavedFilters = filterParams.some(key => !isNil(savedParams[key]));
+
+      // Don't restore the saved filters if there are already filters specified (via url, ui-sref, etc)
+      const shouldRedirectWithSavedParams = !hasFilters && hasSavedFilters;
+      return shouldRedirectWithSavedParams ? trans.targetState().withParams(savedParams) : null;
+    });
+
+    // When switching apps, clear the saved state
+    transitionService.onStart({ exiting: '**.application' }, () => {
```

Comment:
forgive my ignorance — is this `exiting: '**.application'` bit also triggered when going from one app's states to another app's states directly? or only when going from an application state to a non-application state like `search`?